04-cs-fundamentals/internship/graphs.md

```diff
@@ -0,0 +1,421 @@
+
+  # shortest_distances will hold the shortest distances from start_node to i
+  # it starts with infinity as the value
+  shortest_distances = Array.new(num_nodes, Float::INFINITY)
+
+
+  # added[i] will be true if the path to i
+  # from the source has been found
+  added = Array.new(num_nodes, false)
+
+
+  # Distance of source vertex from
+  # itself is always 0
+  shortest_distances[start_node] = 0
+  
+  # parent array to store the shortest path tree
+  parents = Array.new(num_nodes)
+  # no parent for the start node
+  parents[start_node] = nil
+
+
+  # Find shortest path for all nodes
+  (num_nodes - 1).times do
+
+    # Pick the minimum distance vertex
+    # from the set of vertices not yet
+    # processed. nearest_node is  
+    # always equal to start_node in  
+    # first iteration.
+    nearest_node = -1
+    shortest_distance = Float::INFINITY
+    (0...num_nodes).each do |node_index|
+      if (!added[node_index] &&
+          shortest_distances[node_index] <  shortest_distance)  
+        nearest_node  = node_index
+        shortest_distance = shortest_distances[node_index]
+      end
+    end  
+
+    # Mark the picked vertex as visited
+    added[nearest_node] = true
+    # Update dist value of the
+    # adjacent nodes of the picked node.
+    (0...num_nodes).each do |node_index|
+      edge_distance = adjacency_matrix[nearest_node][node_index]
+
+      if (edge_distance > 0 && 
+          ((shortest_distance + edge_distance) <  
+              shortest_distances[node_index]))  
+        parents[node_index] = nearest_node
+        shortest_distances[node_index] = shortest_distance +  
+                                        edge_distance
+      end
+    end
+  end
+  return {  
+    start_node: start_node,
+    parent_list: parents,
+    shortest_distances: shortest_distances
+  }
+end
+```
+
+**Questions**
+
+<details>
+  <summary>Is Dijkstra's Algorithm a greedy algorithm?</summary>
+  <strong>YES!</strong>  With each iteration of the main loop Dijkstra's algorithm picks the next node with the shortest path.  It makes each choice by picking the next shortest path available.  
+</details>
+
+<details>
+  <summary>What is the time complexity of Dijkstra's Algorithm?</summary>
+  <strong>O(N<sup>2</sup>)</strong>  Where N is the number of nodes.  The algorithm iterates until each node has been visited, N times, and with each iteration you have to find the node with the smallest current distance, so with each iteration you have to look roughly at most N other nodes.
+</details>
+
+**Exercise** 
+
+Take a look at the solution above.  Can you convert it to a solution involving an Adjacency list?  Each entry in the adjacency list will need to store a weight as well as the destination node.
+
+
+#### Using a Heap
+
+Because finding an unvisited node with a minimum distance takes O(N) time, Dijkstra's algorithm can be made more efficient by using a Min-Heap where nodes are stored by order of their distance.  
```

Comment:
```suggestion
Because finding an unvisited node with a minimum distance takes _O(N)_ time, Dijkstra's algorithm can be made more efficient by using a Min-Heap where nodes are stored by order of their distance.  
```